### DISCUSSION TITLE
Possible C# Compiler bug RE Constraints

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/roslyn#55620] [Q&A] [answered]
The code below seems to me that it should compile but it does not.
Here's what seems wonky:
```MyBase``` specifies a constraint on ```Parse```
```MyClass``` implements the ```Parse``` function
You get a compiler error saying you can't specify constraints (other than class or struct) if you uncomment ```where T : DisplayRecord```
You get a compiler errors if no constraints are specified on ```MyClass```'s ```Parse```.
The code only compiles if you specify ```where T : class``` on ```MyClass``` but this seems as though it shouldn't be necessary because ```T : DisplayRecord``` and ```DisplayRecord : record``` and ```record : class```.

Thoughts?

```
#nullable enable

using System;

namespace ConsoleApp1
{
    
    public class MyClass : MyBase
    {
        public override T? Parse<T>(string Text)
            //where T : DisplayRecord
            //where T: class

        {
            return base.Parse<T>(Text);
        }
    }

    public abstract class MyBase
    {
        public virtual T? Parse<T>(string Text) 
            where T : DisplayRecord
        {
            throw new NotImplementedException();
        }
    }

    public record DisplayRecord
    {

    }

}

```

**Comment by svick** [comment, marked as answer]
While it's definitely confusing, I think this is the expected behavior. The reasoning is explained in [the spec for the related C# 9.0 `default` constraint](https://github.com/dotnet/csharplang/blob/134702a/proposals/csharp-9.0/unconstrained-type-parameter-annotations.md#default-constraint):

> For compatibility with existing code where overridden [...] generic methods could not include explicit constraint clauses, `T?` in an overridden [...] method is treated as `Nullable<T>` where `T` is a value type.
>
> To allow annotations for type parameters constrained to reference types, C#8 allowed explicit `where T : class` and `where T : struct` constraints on the overridden [...] method.

**Reply by jnm2** [reply]
Also see the docs page: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/generics/constraints-on-type-parameters#default-constraint

**Comment by TonyValenti** [comment]
Thanks @svick and @jnm2 .  Adding ```where T : class``` to my derived classes is easy enough but it seems as though C# should be smart enough to infer that from the base class.  It feels as though I'm re-specifying information the compiler already has.

If you can pass this up the chain as a feature request, that'd be great.  If not, ```where T : class``` isn't that difficult to add!

**Reply by jnm2** [reply]
Well, I'm curious, so: https://github.com/dotnet/csharplang/discussions/5055

**Reply by jnm2** [reply]
That was quick! @svick shows here why inferring `where T : class` would break backward compatibility with C# 9: https://github.com/dotnet/csharplang/discussions/5055#discussioncomment-1185532
